sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuPythonHelper.scala

```diff
@@ -0,0 +1,105 @@
+import com.nvidia.spark.rapids.python.PythonConfEntries._
+
+import org.apache.spark.SparkEnv
+import org.apache.spark.api.python.ChainedPythonFunctions
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config.{CPUS_PER_TASK, EXECUTOR_CORES}
+import org.apache.spark.internal.config.Python.{PYTHON_USE_DAEMON, PYTHON_WORKER_MODULE}
+
+object GpuPythonHelper extends Logging {
+
+  private lazy val rapidsConf = new RapidsConf(SparkEnv.get.conf)
+  private lazy val gpuId = GpuDeviceManager.getDeviceId()
+    .getOrElse(throw new IllegalStateException("No gpu id!"))
+    .toString
+  private lazy val isPythonPooledMemEnabled = rapidsConf.get(PYTHON_POOLED_MEM)
+    .getOrElse(rapidsConf.isPooledMemEnabled)
+    .toString
+  private lazy val isPythonUvmEnabled = rapidsConf.get(PYTHON_UVM_ENABLED)
+    .getOrElse(rapidsConf.isUvmEnabled)
+    .toString
+  private lazy val (initAllocPerWorker, maxAllocPerWorker) = {
+    val info = Cuda.memGetInfo()
+    val maxFactionTotal = rapidsConf.get(PYTHON_RMM_MAX_ALLOC_FRACTION)
+    val maxAllocTotal = (maxFactionTotal * info.total).toLong
+    // Initialize pool size for all pythons workers. If the fraction is not set,
+    // use half of the free memory as default.
+    val initAllocTotal = rapidsConf.get(PYTHON_RMM_ALLOC_FRACTION)
+      .map { fraction =>
+        if (0 < maxFactionTotal && maxFactionTotal < fraction) {
+          throw new IllegalArgumentException(s"The value of '$PYTHON_RMM_MAX_ALLOC_FRACTION' " +
+            s"should not be less than that of '$PYTHON_RMM_ALLOC_FRACTION', but found " +
+            s"$maxFactionTotal < $fraction")
+        }
+        (fraction * info.total).toLong
+      }
+      .getOrElse((0.5 * info.free).toLong)
+    if (initAllocTotal > info.free) {
+      logWarning(s"Initial RMM allocation(${initAllocTotal / 1024.0 / 1024} MB) for " +
+        s"all the Python workers is larger than free memory(${info.free / 1024.0 / 1024} MB)")
+    } else {
+      logDebug(s"Configure ${initAllocTotal / 1024.0 / 1024}MB GPU memory for " +
+        s"all the Python workers.")
+    }
+
+    // Calculate the pool size for each Python worker.
+    val concurrentPythonWorkers = rapidsConf.get(CONCURRENT_PYTHON_WORKERS)
+    val sparkConf = SparkEnv.get.conf
+    // Spark does not throw exception even the value of CPUS_PER_TASK is negative, so
+    // return 1 if it is less than zero to continue the task.
+    val cpuTaskSlots = sparkConf.get(EXECUTOR_CORES) / Math.max(1, sparkConf.get(CPUS_PER_TASK))
+    if (0 < concurrentPythonWorkers && concurrentPythonWorkers <= cpuTaskSlots) {
+      (initAllocTotal / concurrentPythonWorkers, maxAllocTotal / concurrentPythonWorkers)
+    } else {
+      // When semaphore is disabled or invalid, use the number of cpu task slots instead.
+      (initAllocTotal / cpuTaskSlots, maxAllocTotal / cpuTaskSlots)
+    }
+  }
+
+  def injectGpuInfo(funcs: Seq[ChainedPythonFunctions]): Unit = {
```

Comment:
nit : I would rather have us pass in the sparkConf instead of getting the global one, but this is likely fairly minor.

Reply:
sparkConf is also used to initialize most of the members in object `GpuPythonHelper`, besides in this function.

For this case i just unified it as a member too to avoid duplicate calls to `SparkEnv.get.conf`. 

Another solution is to add a new constructor with a SparkConf as the argument to require passing a sparkConf when creating a GpuPythonHelper instance.

Reply:
The main issue is that all of these configs are now set in stone.  I cannot dynamically set them as a part of the same spark context.

Reply:
Is it for integration tests or other cases to dynamically set them ?

Is it OK to set them here only when not specified ?

Reply:
Anyway i moved the checking and setting to driver side and the configs are set only when not specified.
Please let me know if this could resolve your concern.